File: javascript/packages/highlighter/src/diagnostic-renderer.ts
import { colorize, hyperlink, severityColor } from "./color.js"
import { visibleWidth, ANSI_REGEX_START, ANSI_ESCAPE } from "./ansi.js"
import { dimStyledText } from "./util.js"
import { LineWrapper } from "./line-wrapper.js"
import * as gutter from "./gutter.js"
import { computeDiagnosticMarkers } from "./diagnostic-markers.js"

import type { DiagnosticMarker } from "./diagnostic-markers.js"
import type { SyntaxRenderer } from "./syntax-renderer.js"
import type { Diagnostic } from "@herb-tools/core"

export interface DiagnosticRenderOptions {
  contextLines?: number
  showLineNumbers?: boolean
  optimizeHighlighting?: boolean
  wrapLines?: boolean
  maxWidth?: number
  truncateLines?: boolean
  codeUrl?: string
  fileUrl?: string
  suffix?: string
}

export class DiagnosticRenderer {
  private syntaxRenderer: SyntaxRenderer

  constructor(syntaxRenderer: SyntaxRenderer) {
    this.syntaxRenderer = syntaxRenderer
  }

  private highlightBackticks(text: string): string {
    if (process.stdout.isTTY && process.env.NO_COLOR === undefined) {
      const boldWhite = "\x1b[1m\x1b[37m"
      const reset = "\x1b[0m"
      return text.replace(/`([^`]+)`/g, `${boldWhite}$1${reset}`)
    }

    return text
  }

  private truncateLineForDiagnostic(
    line: string,
    diagnosticStart: number,
    diagnosticEnd: number,
    maxWidth: number
  ): { line: string; adjustedStart: number; adjustedEnd: number } {
    const plainLineLength = visibleWidth(line)

    if (plainLineLength <= maxWidth) {
      return { line, adjustedStart: diagnosticStart, adjustedEnd: diagnosticEnd }
    }

    const ellipsisChar = "…"
    const ellipsis = colorize(ellipsisChar, "dim")
    const rightPadding = 2
    const ellipsisCharLength = ellipsisChar.length
    const ellipsisLength = ellipsisCharLength + rightPadding

    if (diagnosticStart < maxWidth / 3) {
      const availableWidth = maxWidth - ellipsisLength
      const truncated = LineWrapper.truncateLine(line, availableWidth)
      return {
        line: truncated,
        adjustedStart: diagnosticStart,
        adjustedEnd: Math.min(diagnosticEnd, availableWidth)
      }
    }

    if (diagnosticStart > plainLineLength - maxWidth / 3) {
      const availableWidth = maxWidth - ellipsisLength
      const startPos = Math.max(0, plainLineLength - availableWidth)

      const visiblePortion = this.extractPortionFromPosition(line, startPos, plainLineLength)
      const truncated = ellipsis + visiblePortion

      return {
        line: truncated,
        adjustedStart: Math.max(0, diagnosticStart - startPos + ellipsisCharLength),
        adjustedEnd: Math.max(0, diagnosticEnd - startPos + ellipsisCharLength)
      }
    }

    const contextWidth = maxWidth - (ellipsisLength * 2)
    const contextStart = Math.max(0, diagnosticStart - contextWidth / 3)
    const contextEnd = Math.min(plainLineLength, contextStart + contextWidth)

    const visiblePortion = this.extractPortionFromPosition(line, contextStart, contextEnd)
    const truncated = ellipsis + visiblePortion + ellipsis

    return {
      line: truncated,
      adjustedStart: diagnosticStart - contextStart + ellipsisCharLength,
      adjustedEnd: diagnosticEnd - contextStart + ellipsisCharLength
    }
  }

  private extractPortionFromPosition(styledLine: string, startPos: number, endPos: number): string {
    let styledIndex = 0
    let plainIndex = 0
    let result = ""
    let inRange = false

    while (styledIndex < styledLine.length && plainIndex <= endPos) {
      const char = styledLine[styledIndex]

      if (char === ANSI_ESCAPE) {
        const ansiMatch = styledLine.slice(styledIndex).match(ANSI_REGEX_START)
        if (ansiMatch) {
          if (inRange || plainIndex >= startPos) {
            result += ansiMatch[0]
          }
          styledIndex += ansiMatch[0].length
          continue
        }
      }

      if (plainIndex >= startPos && !inRange) {
        inRange = true
      }

      if (inRange) {
        result += char
      }

      styledIndex++
      plainIndex++
    }

    return result
  }

  renderSingle(
    path: string,
    diagnostic: Diagnostic,
    content: string,
    options: DiagnosticRenderOptions = {},
  ): string {
    const {
      contextLines = 2,
      showLineNumbers = true,
      optimizeHighlighting = true,
      wrapLines = true,
      maxWidth = LineWrapper.getTerminalWidth(),
      truncateLines = false,
    } = options

    const shouldWrap = wrapLines && !truncateLines
    const shouldTruncate = truncateLines
    const fileHeaderText = `${colorize(path, "cyan")}:${colorize(`${diagnostic.location.start.line}:${diagnostic.location.start.column}`, "cyan")}`
    const { codeUrl, fileUrl: fileUrlOption } = options
    const fileHeader = fileUrlOption ? hyperlink(fileHeaderText, fileUrlOption) : fileHeaderText

    const color = severityColor(diagnostic.severity)
    const text = colorize(colorize(diagnostic.severity, color), "bold")
    const diagnosticIdText = diagnostic.code || "-"
    const diagnosticId = codeUrl ? hyperlink(diagnosticIdText, codeUrl) : diagnosticIdText

    const originalLines = content.split("\n")

    const markers = computeDiagnosticMarkers(diagnostic.location, originalLines)
    const markersByLine = new Map<number, DiagnosticMarker>(markers.map(marker => [marker.line, marker]))

    const firstMarkedLine = markers[0].line
    const lastMarkedLine = markers[markers.length - 1].line

    const startLine = Math.max(1, firstMarkedLine - contextLines)
    const endLine = Math.min(originalLines.length, lastMarkedLine + contextLines)

    let lines: string[]
    let lineOffset = 0

    if (optimizeHighlighting) {
      const relevantLines = []

      for (let i = startLine; i <= endLine; i++) {
        relevantLines.push(originalLines[i - 1] || "")
      }

      const relevantContent = relevantLines.join("\n")
      const highlightedContent = this.syntaxRenderer.highlight(relevantContent)

      lines = highlightedContent.split("\n")
      lineOffset = startLine - 1
    } else {
      const highlightedContent = this.syntaxRenderer.highlight(content)
      lines = highlightedContent.split("\n")
      lineOffset = 0
    }

    const gutterPrefix = showLineNumbers ? gutter.continuationPrefix() : ""
    const contentWidth = showLineNumbers ? gutter.availableWidth(maxWidth) : maxWidth

    let contextOutput = ""

    for (let i = startLine; i <= endLine; i++) {
      const line = lines[i - 1 - lineOffset] || ""
      const marker = markersByLine.get(i)
      const isTargetLine = marker !== undefined

      let markerStart = marker ? marker.start : 0
      let markerLength = marker ? Math.max(1, marker.end - marker.start) : 0

      const prefix = showLineNumbers ? gutter.linePrefix(i, isTargetLine, isTargetLine ? color : undefined) : ""

      const displayLine = isTargetLine ? line : dimStyledText(line)

      if (shouldWrap) {
        const wrappedLines = LineWrapper.wrapLine(displayLine, contentWidth, "")

        for (let j = 0; j < wrappedLines.length; j++) {
          if (j === 0) {
            contextOutput += `${prefix}${wrappedLines[j]}\n`
          } else {
            contextOutput += `${gutterPrefix}${wrappedLines[j]}\n`
          }
        }
      } else if (shouldTruncate) {
        let truncatedLine: string

        if (marker) {
          const result = this.truncateLineForDiagnostic(displayLine, marker.start, marker.end, contentWidth)
          truncatedLine = result.line
          markerStart = result.adjustedStart
          markerLength = Math.max(1, result.adjustedEnd - result.adjustedStart)
        } else {
          truncatedLine = LineWrapper.truncateLine(displayLine, contentWidth)
        }

        contextOutput += `${prefix}${truncatedLine}\n`
      } else {
        contextOutput += `${prefix}${displayLine}\n`
      }

      if (marker) {
        const pointer_prefix = showLineNumbers ? gutter.pointerPrefix() : ""
        const pointerSpacing = " ".repeat(Math.max(0, markerStart + (showLineNumbers ? 1 : 0)))
        const pointer = colorize("~".repeat(markerLength), color)

        contextOutput += `${pointer_prefix}${pointerSpacing}${pointer}\n`
      }
    }

    const highlightedMessage = this.highlightBackticks(diagnostic.message)
    const { suffix } = options
    const suffixText = suffix ? ` ${suffix}` : ""
    const header = showLineNumbers ? `${fileHeader}\n\n` : ""

    return `[${text}] ${highlightedMessage} (${diagnosticId})${suffixText}

${header}${contextOutput.trimEnd()}
`
  }
}

File: rust/herb-highlighter/src/diagnostic_renderer.rs
use std::fmt::Write as _;

use std::collections::HashMap;

use crate::ansi::{ansi_sequence_at_start, visible_width};
use crate::color::{colorize, hyperlink, is_color_enabled, is_terminal, severity_color, NamedColor};
use crate::diagnostic::Diagnostic;
use crate::diagnostic_markers::{compute_diagnostic_markers, DiagnosticMarker};
use crate::gutter;
use crate::line_wrapper::LineWrapper;
use crate::syntax_renderer::SyntaxRenderer;
use crate::text_formatter::replace_backticks;
use crate::util::dim_styled_text;

#[derive(Debug, Clone)]
pub struct DiagnosticRenderOptions {
  pub context_lines: usize,
  pub show_line_numbers: bool,
  pub optimize_highlighting: bool,
  pub wrap_lines: bool,
  pub max_width: Option<usize>,
  pub truncate_lines: bool,
  pub code_url: Option<String>,
  pub file_url: Option<String>,
  pub suffix: Option<String>,
}

impl Default for DiagnosticRenderOptions {
  fn default() -> Self {
    Self {
      context_lines: 2,
      show_line_numbers: true,
      optimize_highlighting: true,
      wrap_lines: true,
      max_width: None,
      truncate_lines: false,
      code_url: None,
      file_url: None,
      suffix: None,
    }
  }
}

struct TruncatedLine {
  line: String,
  adjusted_start: f64,
  adjusted_end: f64,
}

pub struct DiagnosticRenderer<'a> {
  syntax_renderer: &'a SyntaxRenderer,
}

impl<'a> DiagnosticRenderer<'a> {
  pub fn new(syntax_renderer: &'a SyntaxRenderer) -> Self {
    Self { syntax_renderer }
  }

  fn highlight_backticks(&self, text: &str) -> String {
    if is_terminal() && is_color_enabled() {
      return replace_backticks(text, "\x1b[1m\x1b[37m", "\x1b[0m");
    }

    text.to_string()
  }

  fn truncate_line_for_diagnostic(&self, line: &str, diagnostic_start: usize, diagnostic_end: usize, max_width: usize) -> TruncatedLine {
    let plain_line_length = visible_width(line);

    if plain_line_length <= max_width {
      return TruncatedLine {
        line: line.to_string(),
        adjusted_start: diagnostic_start as f64,
        adjusted_end: diagnostic_end as f64,
      };
    }

    let ellipsis_character = "…";
    let ellipsis = colorize(ellipsis_character, NamedColor::Dim);
    let right_padding = 2;
    let ellipsis_character_length = ellipsis_character.chars().count();
    let ellipsis_length = ellipsis_character_length + right_padding;

    if (diagnostic_start as f64) < max_width as f64 / 3.0 {
      let available_width = max_width - ellipsis_length;
      let truncated = LineWrapper::truncate_line(line, available_width);

      return TruncatedLine {
        line: truncated,
        adjusted_start: diagnostic_start as f64,
        adjusted_end: diagnostic_end.min(available_width) as f64,
      };
    }

    if diagnostic_start as f64 > plain_line_length as f64 - max_width as f64 / 3.0 {
      let available_width = max_width - ellipsis_length;
      let start_position = plain_line_length.saturating_sub(available_width);

      let visible_portion = self.extract_portion_from_position(line, start_position as f64, plain_line_length as f64);
      let truncated = format!("{ellipsis}{visible_portion}");

      let offset = start_position as f64 - ellipsis_character_length as f64;

      return TruncatedLine {
        line: truncated,
        adjusted_start: (diagnostic_start as f64 - offset).max(0.0),
        adjusted_end: (diagnostic_end as f64 - offset).max(0.0),
      };
    }

    let context_width = (max_width - ellipsis_length * 2) as f64;
    let context_start = (diagnostic_start as f64 - context_width / 3.0).max(0.0);
    let context_end = (plain_line_length as f64).min(context_start + context_width);

    let visible_portion = self.extract_portion_from_position(line, context_start, context_end);
    let truncated = format!("{ellipsis}{visible_portion}{ellipsis}");

    TruncatedLine {
      line: truncated,
      adjusted_start: diagnostic_start as f64 - context_start + ellipsis_character_length as f64,
      adjusted_end: diagnostic_end as f64 - context_start + ellipsis_character_length as f64,
    }
  }

  fn extract_portion_from_position(&self, styled_line: &str, start_position: f64, end_position: f64) -> String {
    let mut styled_index = 0;
    let mut plain_index = 0.0;
    let mut result = String::new();
    let mut in_range = false;

    while styled_index < styled_line.len() && plain_index <= end_position {
      if let Some(sequence) = ansi_sequence_at_start(&styled_line[styled_index..]) {
        if in_range || plain_index >= start_position {
          result.push_str(sequence);
        }

        styled_index += sequence.len();

        continue;
      }

      let character = styled_line[styled_index..].chars().next().unwrap_or('\0');

      if plain_index >= start_position && !in_range {
        in_range = true;
      }

      if in_range {
        result.push(character);
      }

      styled_index += character.len_utf8();
      plain_index += 1.0;
    }

    result
  }

  pub fn render_single(&self, path: &str, diagnostic: &Diagnostic, content: &str, options: &DiagnosticRenderOptions) -> String {
    let context_lines = options.context_lines;
    let show_line_numbers = options.show_line_numbers;
    let optimize_highlighting = options.optimize_highlighting;
    let max_width = options.max_width.unwrap_or_else(LineWrapper::get_terminal_width);
    let truncate_lines = options.truncate_lines;

    let should_wrap = options.wrap_lines && !truncate_lines;
    let should_truncate = truncate_lines;

    let file_header_text = format!(
      "{}:{}",
      colorize(path, NamedColor::Cyan),
      colorize(
        &format!("{}:{}", diagnostic.location.start.line, diagnostic.location.start.column),
        NamedColor::Cyan
      )
    );

    let file_header = match &options.file_url {
      Some(file_url) => hyperlink(&file_header_text, file_url),
      None => file_header_text,
    };

    let color = severity_color(diagnostic.severity);
    let text = colorize(&colorize(diagnostic.severity.as_str(), color), NamedColor::Bold);
    let diagnostic_id_text = diagnostic.code.clone().unwrap_or_else(|| "-".to_string());

    let diagnostic_id = match &options.code_url {
      Some(code_url) => hyperlink(&diagnostic_id_text, code_url),
      None => diagnostic_id_text,
    };

    let original_lines: Vec<&str> = content.split('\n').collect();

    let markers = compute_diagnostic_markers(&diagnostic.location, &original_lines);
    let markers_by_line: HashMap<usize, DiagnosticMarker> = markers.iter().map(|marker| (marker.line, *marker)).collect();

    let first_marked_line = markers[0].line;
    let last_marked_line = markers[markers.len() - 1].line;

    let start_line = 1.max(first_marked_line.saturating_sub(context_lines));
    let end_line = original_lines.len().min(last_marked_line + context_lines);

    let highlighted_content;
    let lines: Vec<&str>;
    let line_offset;

    if optimize_highlighting {
      let mut relevant_lines: Vec<&str> = Vec::new();

      for number in start_line..=end_line {
        relevant_lines.push(original_lines.get(number - 1).copied().unwrap_or(""));
      }

      let relevant_content = relevant_lines.join("\n");

      highlighted_content = self.syntax_renderer.highlight(&relevant_content);
      lines = highlighted_content.split('\n').collect();
      line_offset = start_line - 1;
    } else {
      highlighted_content = self.syntax_renderer.highlight(content);
      lines = highlighted_content.split('\n').collect();
      line_offset = 0;
    }

    let gutter_prefix = if show_line_numbers { gutter::continuation_prefix() } else { String::new() };
    let available_width = if show_line_numbers { gutter::available_width(max_width) } else { max_width };

    let mut context_output = String::new();

    for number in start_line..=end_line {
      let line = lines.get(number - 1 - line_offset).copied().unwrap_or("");
      let marker = markers_by_line.get(&number);
      let is_target_line = marker.is_some();

      let mut marker_start = marker.map_or(0.0, |marker| marker.start as f64);
      let mut marker_length = marker.map_or(0.0, |marker| (marker.end as f64 - marker.start as f64).max(1.0));

      let line_prefix = if show_line_numbers {
        gutter::line_prefix(number, is_target_line, is_target_line.then_some(color))
      } else {
        String::new()
      };

      let display_line = if is_target_line { line.to_string() } else { dim_styled_text(line) };

      if should_wrap {
        let wrapped_lines = LineWrapper::wrap_line(&display_line, available_width, "");

        for (index, wrapped_line) in wrapped_lines.iter().enumerate() {
          if index == 0 {
            let _ = writeln!(context_output, "{line_prefix}{wrapped_line}");
          } else {
            let _ = writeln!(context_output, "{gutter_prefix}{wrapped_line}");
          }
        }
      } else if should_truncate {
        let truncated_line = match marker {
          Some(marker) => {
            let result = self.truncate_line_for_diagnostic(&display_line, marker.start, marker.end, available_width);

            marker_start = result.adjusted_start;
            marker_length = (result.adjusted_end - result.adjusted_start).max(1.0);

            result.line
          }

          None => LineWrapper::truncate_line(&display_line, available_width),
        };

        let _ = writeln!(context_output, "{line_prefix}{truncated_line}");
      } else {
        let _ = writeln!(context_output, "{line_prefix}{display_line}");
      }

      if marker.is_some() {
        let pointer_prefix = if show_line_numbers { gutter::pointer_prefix() } else { String::new() };
        let spacing = (marker_start + if show_line_numbers { 1.0 } else { 0.0 }).max(0.0) as usize;
        let pointer_spacing = " ".repeat(spacing);
        let pointer = colorize(&"~".repeat(marker_length as usize), color);

        let _ = writeln!(context_output, "{pointer_prefix}{pointer_spacing}{pointer}");
      }
    }

    let highlighted_message = self.highlight_backticks(&diagnostic.message);
    let suffix_text = match &options.suffix {
      Some(suffix) => format!(" {suffix}"),
      None => String::new(),
    };
    let header = if show_line_numbers { format!("{file_header}\n\n") } else { String::new() };

    format!(
      "[{text}] {highlighted_message} ({diagnostic_id}){suffix_text}\n\n{header}{}\n",
      context_output.trim_end()
    )
  }
}
